src/Http/Cookie/Cookie.php

```diff
@@ -0,0 +1,531 @@
+     *
+     * @link http://php.net/manual/en/function.setcookie.php
+     * @param string $name Cookie name
+     * @param string|array $value Value of the cookie
+     * @param \DateTimeInterface|null $expiresAt Expiration time and date
+     * @param string $path Path
+     * @param string $domain Domain
+     * @param bool $secure Is secure
+     * @param bool $httpOnly HTTP Only
+     */
+    public function __construct($name, $value = '', $expiresAt = null, $path = '', $domain = '', $secure = false, $httpOnly = false)
+    {
+        $this->validateName($name);
+        $this->setName($name);
+        $this->setValue($value);
+        $this->setDomain($domain);
+        $this->setHttpOnly($httpOnly);
+        $this->setPath($path);
+
+        if ($expiresAt !== null) {
+            $this->expiresAt($expiresAt);
+        }
+    }
+
+    /**
+     * Builds the expiration value part of the header string
+     *
+     * @return string
+     */
+    protected function _buildExpirationValue()
+    {
+        return sprintf(
+            'expires=%s',
+            gmdate('D, d-M-Y H:i:s T', $this->expiresAt)
+        );
+    }
+
+    /**
+     * Returns a header value as string
+     *
+     * @return string
+     */
+    public function toHeaderValue()
+    {
+        $headerValue[] = sprintf('%s=%s', $this->name, urlencode($this->value));
+
+        if ($this->expiresAt !== 0) {
+            $headerValue[] = $this->_buildExpirationValue();
+        }
+        if ($this->path !== '') {
+            $headerValue[] = sprintf('path=%s', $this->path);
+        }
+        if ($this->domain !== '') {
+            $headerValue[] = sprintf('domain=%s', $this->domain);
+        }
+        if ($this->secure) {
+            $headerValue[] = 'secure';
+        }
+        if ($this->httpOnly) {
+            $headerValue[] = 'httponly';
+        }
+
+        return implode('; ', $headerValue);
+    }
+
+    /**
+     * Sets the cookie name
+     *
+     * @param string $name Name of the cookie
+     * @return $this
+     */
+    public function setName($name)
+    {
+        $this->validateName($name);
+        $this->name = $name;
+
+        return $this;
+    }
+
+    /**
+     * Gets the cookie name
+     *
+     * @return string
+     */
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    /**
+     * Validates the cookie name
+     *
+     * @param string $name Name of the cookie
+     * @return void
+     * @throws \InvalidArgumentException
+     */
+    protected function validateName($name)
+    {
+        if (preg_match("/[=,; \t\r\n\013\014]/", $name)) {
+            throw new InvalidArgumentException(
+                sprintf('The cookie name `%s` contains invalid characters.', $name)
+            );
+        }
+
+        if (empty($name)) {
+            throw new InvalidArgumentException('The cookie name cannot be empty.');
+        }
+    }
+
+    /**
+     * Gets the raw cookie value
```

Comment:
After calling `expand()` this is no longer "raw" value?
Also this method may return `array` in that case, am I right?